### ISSUE TITLE
Migration: type enum_Notes_permission already exists

### ISSUE THREAD

**Issue by haslersn** [hedgedoc/hedgedoc#892]
I tried to migrate from CodiMD a160d81fe33044ca8fbb71addd77c40d55b37251 to HedgeDoc 1.7.2 using the docker image `quay.io/hedgedoc/hedgedoc:1.7.2-alpine`. I run the migration using the entrypoint

```
- node_modules/.bin/sequelize
- db:migrate
- --config
- /hedgedoc-config/db.js
```

where `/hedgedoc-config/db.js` has the following content:

```
module.exports = {
    "production": {
        "dialect": "postgres",
        "host": "hedgedoc-db",
        "username": "user",
        "password": process.env.DB_PASSWORD,
        "database": "db",
        "dialectOptions": {
            "ssl": {
                "require": true,
                "rejectUnauthorized": false
            }
        }
    }
}
```

Unfortunately, the migration throws the following error:

```
Sequelize CLI [Node: 12.20.1, CLI: 5.5.1, ORM: 5.22.3]

Loaded configuration file "../hedgedoc-config/db.js".
Using environment "production".
== 20180306150303-fix-enum: migrating =======

ERROR: type "enum_Notes_permission" already exists
```

When I choose to ignore the error, HedgeDoc seems to work just fine.

**Expected behavior**

The migration should not fail and should return with code 0.

**Additional context**

I'm deploying HedgeDoc on Kubernetes using the following Deployment: https://gist.github.com/haslersn/ef6f880fa04fa827bd3ad5c11ed63a33

**Comment by SISheogorath** [comment]
I don't think it's healthy to ignore db migration errors: https://gist.github.com/haslersn/ef6f880fa04fa827bd3ad5c11ed63a33#file-hedgedoc-deployment-yaml-L26

There is currently still a little problem with how migrations are implemented in 1.x, not only that they are in an own tool, but also that the main app pushes its database schema by itself. when the latter happens, the migrations will no longer work successful, as you experience here.

we should backport this PR to 1.x: https://github.com/hedgedoc/hedgedoc/pull/384 and solve those issues once and for all.

**Comment by haslersn** [comment]
Is there some manual way to fix it in the database, such that migrations can be successful again?

**Comment by SISheogorath** [comment]
Yes, you can add the filenames to a sequel items radars table. You should find some examples in there already.

**Comment by fefrei** [comment]
I'm having the same issue migrating from `quay.io/codimd/server:1.5.0` to `quay.io/hedgedoc/hedgedoc:latest` (not deployed on Kubernetes, just a local Docker Compose setup), I had to restore to a backup.

> Yes, you can add the filenames to a sequel items radars table

I don't understand that, sorry. Anything I can do to migrate the database?

**Comment by SISheogorath** [comment]
Sorry, my comment went through my phones autocorrelation which always messes up table names or technical details. I meant `sequelize-meta`-table.

**Comment by fefrei** [comment]
Hm, I'm still not entirely sure what you mean. (I'm familiar with databases, but have never worked with `sequelize` directly.)
Are you suggesting editing the database to convince `sequelize` this particular migration has already been run?

**Comment by fefrei** [comment]
(Taking your 👍 as a _Yes_:) Do you have any hints on how to do so?
Either way, this sounds really fishy. I don't understand why this migration needs to be run anyways – it looks like it adds the _limited_ and _protected_ permissions for notes, which I've always had (and have used successfully) since my initial installation.

I should also note that I did not rename the database tables and users to `hedgedoc` – they're still on the old defaults from 1.5.0 days, maybe that's related somehow.

**Comment by DerMolly** [comment]
> Do you have any hints on how to do so?

As @SISheogorath already wrote: There is a table called `sequelize-meta`, where already run migrations get added. You should be able to add the one that gives you trouble there. You should find some examples in there already.

> I don't understand why this migration needs to be run anyways

That's how `sequelize` or rather any ORM  or QueryBuilder keeps track of the DB state. Any change we make to DB structure needs to be done via a migration. So the system can decide which code to run to get the DB in the state it's expecting.

> I should also note that I did not rename the database tables and users to `hedgedoc` – they're still on the old defaults from 1.5.0 days, maybe that's related somehow.

As @SISheogorath wrote:

> There is currently still a little problem with how migrations are implemented in 1.x, not only that they are in an own tool, but also that the main app pushes its database schema by itself. **when the latter happens, the migrations will no longer work successful, as you experience here.**

You can ignore the migration errors of course if you like, but we wouldn't recommend it.

Yes we're aware of the problematic migration situation and are trying to fix that for 2.0, but not much we can do at this point without rebuilding/restructuring parts of 1.7.2.

**Comment by fefrei** [comment]
Thanks for clarifying!

> That's how sequelize or rather any ORM or QueryBuilder keeps track of the DB state. Any change we make to DB structure needs to be done via a migration. So the system can decide which code to run to get the DB in the state it's expecting.

Sorry for being unclear – I didn't get why this _specific_ migration needs running, as it looks like it already ran. That's of course explained by...

> the main app pushes its database schema by itself. when the latter happens, the migrations will no longer work successful, as you experience here

...which I failed to read properly.

Based on all this changing for 2.0, I'll likely stick with 1.5 and hope for the upgrade to 2.0 to work fine 🙂

**Comment by DerMolly** [comment]
> Based on all this changing for 2.0, I'll likely stick with 1.5 and hope for the upgrade to 2.0 to work fine :slightly_smiling_face:

I'd strongly advice you to update to 1.7.2, because there were some critical security issues we fixed since 1.5.0 @fefrei

**Comment by fefrei** [comment]
Thanks for insisting 🙂

I got it to work. Here's some notes on what I did, in case someone has the same issue:

- attempt to start the app as usual, look at the logs to identify the offending migration
- take it `down`
- start just the database: `docker-compose up -d database`
- get into an SQL prompt: `docker-compose exec database psql -Upostgres hedgedoc`
- cheat Sequqlize into thinking the migration is done: `INSERT INTO "SequelizeMeta" (name) VALUES ('20180306150303-fix-enum.js');`
- disconnect, stop the container, start up as usual

The following startup still had some migration weirdness...

```
database_1  |
database_1  | PostgreSQL Database directory appears to contain a database; Skipping initialization
database_1  |
database_1  | LOG:  database system was shut down at 2021-02-28 18:43:31 UTC
database_1  | LOG:  MultiXact member wraparound protections are now enabled
database_1  | LOG:  database system is ready to accept connections
database_1  | LOG:  autovacuum launcher started
app_1       | 2021/02/28 18:44:05 Waiting for: tcp://database:5432
app_1       | 2021/02/28 18:44:05 Connected to tcp://database:5432
database_1  | LOG:  incomplete startup packet
app_1       |
app_1       | Sequelize CLI [Node: 12.21.0, CLI: 5.5.1, ORM: 5.22.3]
app_1       |
app_1       | Parsed url postgres://hedgedoc:*****@database:5432/hedgedoc
app_1       | == 20180326103000-use-text-in-tokens: migrating =======
app_1       | == 20180326103000-use-text-in-tokens: migrated (0.410s)
app_1       |
app_1       | == 20180525153000-user-add-delete-token: migrating =======
database_1  | ERROR:  column "deleteToken" of relation "Users" already exists
database_1  | STATEMENT:  ALTER TABLE "public"."Users" ADD COLUMN "deleteToken" UUID;
app_1       | Migration has already run… ignoring.
app_1       | == 20180525153000-user-add-delete-token: migrated (0.121s)
app_1       |
app_1       | == 20200321153000-fix-account-deletion: migrating =======
app_1       | == 20200321153000-fix-account-deletion: migrated (5.568s)
app_1       |
app_1       |
app_1       | > HedgeDoc@1.7.2 start /hedgedoc
app_1       | > sequelize db:migrate && node app.js
app_1       |
app_1       |
app_1       | Sequelize CLI [Node: 12.21.0, CLI: 5.5.1, ORM: 5.22.3]
app_1       |
app_1       | Parsed url postgres://hedgedoc:*****@database:5432/hedgedoc
app_1       | No migrations were executed, database schema was already up to date.
```

...but it all started up fine.

Thank you, @SISheogorath and @DerMolly!

**Comment by DerMolly** [comment]
I guess this can be closed now
